controllers/ipaddress_controller.go

```diff
@@ -0,0 +1,128 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/networking"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/scope"
+	ipamutils "sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/ipam"
+)
+
+// IPAddressReconciler reconciles a IPAddress object.
+type IPAddressReconciler struct {
+	Client           client.Client
+	Recorder         record.EventRecorder
+	WatchFilterValue string
+	ScopeFactory     scope.Factory
+	CaCertificates   []byte // PEM encoded ca certificates.
+
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=ipam.cluster.x-k8x.io.cluster.x-k8s.io,resources=ipaddresses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ipam.cluster.x-k8x.io.cluster.x-k8s.io,resources=ipaddresses/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the IPAddress object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *IPAddressReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("ipaddress", req.NamespacedName)
+
+	ipAddress := &ipamv1.IPAddress{}
+	if err := r.Client.Get(ctx, req.NamespacedName, ipAddress); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if ipAddress.Spec.PoolRef.Kind != openStackFloatingIPPool {
+		log.Info("IPAddress is not associated with OpenStackFloatingIPPool")
+		return ctrl.Result{}, nil
+	}
+
+	pool := &infrav1.OpenStackFloatingIPPool{}
+
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: ipAddress.Namespace,
+		Name:      ipAddress.Spec.PoolRef.Name,
+	}, pool); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log = log.WithValues("openStackFloatingIPPool", pool.Name)
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !ipAddress.ObjectMeta.DeletionTimestamp.IsZero() {
+		// If the IPAddress has more than one finalizer, it has not been released by machine yet and we should not delete it.
+		if controllerutil.ContainsFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer) && len(ipAddress.GetFinalizers()) == 1 {
```

Comment:
I'm not 100% sure about how that use case would look and it might only be the comment that needs to be changed. My reasoning was that if there's another finalizer on the address it's probably being used or something wants to keep it and does not want it deleted from openstack thus waiting for the "DeleteFloatingIPFinalizer" to be the last finalizer left feels reasonable.
As we are trying to keep the infra provider and ipam provider as separate as possible and follow the api spec I feel like expecting a finalizer thats not managed by the ipam provider would counteract those things

Reply:
Okay, I see now, cool!